core/types/transaction_signing.go

```diff
@@ -260,7 +267,22 @@ func (s eip2930Signer) Sender(tx *Transaction) (common.Address, error) {
 		if !tx.Protected() {
 			return HomesteadSigner{}.Sender(tx)
 		}
-		V = new(big.Int).Sub(V, s.chainIdMul)
+		// check if the chain has different chainId from the allowed list
+		if s.chainIdMul == nil {
+			return common.Address{}, ErrInvalidChain
+		}
+
+		var index int
+		var isMainnet, isTestnet bool
+		chainId := tx.ChainId()
+		index, isMainnet = getChainIdIndexAndCheckValid(mainnetValidChains, chainId)
+		if !isMainnet {
+			index, isTestnet = getChainIdIndexAndCheckValid(testnetValidChains, chainId)
+		}
+		if !isMainnet && !isTestnet {
+			return common.Address{}, ErrTxTypeNotSupported
+		}
+		V = new(big.Int).Sub(V, s.chainIdMul[index])
```

Comment:
I think this works satisfies the requirements (allow external txs without enabling replays), but I wonder if all of this additional logic is cheaper than the bigint multiply we saved?

I think you can do away with the search functions, and dual mainnet/testnet lists, and just store the list of `chainIdMuls` as a map in `s`. Then lines 270-285, 403-417, & 431-445 would just reduce to the following short snippet:
```
chainIdMul, found := s.chainIdMuls[chainId]
if !found {
    return common.Address{}, ErrInvalidChain
}
V = new(big.Int).Sub(V, chainIdMul)
```

This would achieve the performance tradeoff we were looking for by precomputing the bigint mul, while still satisfying the two requirements.

Reply:
What you have today works though, and I'd be fine with shipping it as-is and logging an issue to revisit this if performance becomes a priority.